pkg/config/expression.go

```diff
@@ -64,47 +67,216 @@ func SimpleVarToReference(s string) (Reference, error) {
+func TraversalToReference(t hcl.Traversal) (Reference, error) {
+	if t.IsRelative() {
+		return Reference{}, fmt.Errorf("got relative traversal")
+	}
+	getAttrName := func(i int) (string, error) {
+		if i >= len(t) {
+			return "", fmt.Errorf("traversal does not have enough components")
+		}
+		if a, ok := t[i].(hcl.TraverseAttr); ok {
+			return a.Name, nil
+		}
+		return "", fmt.Errorf("got unexpected traversal component: %#v", t[i])
+	}
+	switch root := t.RootName(); root {
+	case "var":
+		n, err := getAttrName(1)
+		if err != nil {
+			return Reference{}, fmt.Errorf("expected second component of global var reference to be a variable name, got %w", err)
+		}
+		return Reference{GlobalVar: true, Name: n}, nil
+	case "module":
+		m, err := getAttrName(1)
+		if err != nil {
+			return Reference{}, fmt.Errorf("expected second component of module var reference to be a module name, got %w", err)
+		}
+		n, err := getAttrName(2)
+		if err != nil {
+			return Reference{}, fmt.Errorf("expected third component of module var reference to be a variable name, got %w", err)
+		}
+		return Reference{Module: m, Name: n}, nil
+	default:
+		return Reference{}, fmt.Errorf("unexpected first component of reference: %#v", root)
+	}
+}
+
+// IsRawHclLiteral checks if passed value of type cty.String
+// and its content starts with "((" and ends with "))".
+// Returns trimmed string and result of test.
+func IsRawHclLiteral(v cty.Value) (string, bool) {
+	if v.Type() != cty.String {
+		return "", false
+	}
+	s := v.AsString()
+	if len(s) < 4 || s[:2] != "((" || s[len(s)-2:] != "))" {
+		return "", false
+	}
+	return s[2 : len(s)-2], true
+}
+
+// HclExpression is a representation of HCL literals in Blueprint
+type HclExpression struct {
+	e  hclsyntax.Expression
+	s  string
+	rs []Reference
+}
+
+// ParseExpression returns HclExpression
+func ParseExpression(s string) (HclExpression, error) {
+	e, diag := hclsyntax.ParseExpression([]byte(s), "", hcl.Pos{})
+	if diag.HasErrors() {
+		return HclExpression{}, diag
+	}
+	ts := e.Variables()
+	rs := make([]Reference, len(ts))
+	for i, t := range ts {
+		var err error
+		if rs[i], err = TraversalToReference(t); err != nil {
+			return HclExpression{}, err
+		}
+	}
+	return HclExpression{e: e, s: s, rs: rs}, nil
+}
+
+// MustParseExpression is "errorless" version of ParseExpression
+// NOTE: only use it if passed expression is guaranteed to be correct
+func MustParseExpression(s string) HclExpression {
+	if exp, err := ParseExpression(s); err != nil {
+		panic(fmt.Errorf("error while parsing %#v: %w", s, err))
+	} else {
+		return exp
+	}
+}
+
+// Tokenize returns Tokens to be used for marshalling HCL
+func (e HclExpression) Tokenize() hclwrite.Tokens {
+	return hclwrite.TokensForIdentifier(e.s)
+}
+
+// References return Reference for all variables used in the expression
+func (e HclExpression) References() []Reference {
+	c := make([]Reference, len(e.rs))
```

Comment:
could we just return `slices.Clone(e.rs)`?

I realize we might get satisfaction from using `make`, but I don't think we are setting capacity here, so I bet behavior is the same.

Reply:
> I don't think we are setting capacity here
We do set capacity, second argument is `len(e.rs)` is exact capacity we want.

> just return `slices.Clone`
I'm hesitant to take dependency on `x/exp` packages with no good reason.